### ISSUE TITLE
New PerlinNoiseAV test in SMHasher

### ISSUE THREAD

**Issue by avaneev** [Cyan4973/xxHash#738]
Sorry to write this directly as I'm an "interested person" having my own hash functions, but recently the claim "xxhash passes all SMHasher tests" became imprecise: I've invented a new Perlin Noise test that finds "holes" in several hashes, not just xxhash. Maybe you can fix this as xxhash is too popular to miss that fact. I'm not expecting to gain anything from my post; my `komihash` hash function is relatively new, so needs a lot more time to gain traction in this very competitive "hash function" field.

Maybe you can also add `komihash` to your tests, it's probably not as fast as xxhash3, though, but useful as a comparison to the "state-of-the-art" hash functions. (`wyhash` comparison would be great as well)

https://github.com/avaneev/komihash
https://github.com/rurban/smhasher

**Comment by Cyan4973** [comment]
Could you point at the Perlin Noise test you refer to ?
The only one I could find in rurban's `smhasher` is : https://github.com/rurban/smhasher/commit/64b7d10ea92701e707ecba5f6b0bacbdb3d98b8a

**Comment by t-mat** [comment]
I'm not sure but this issue may suggest the following new result of 'PerlinNoiseAV' test:

https://github.com/rurban/smhasher/commit/9ca488454eda4dae9ac795147696135e5cdf7dbe

In these new results, `'Perlin Noise'` test is traditional perlin noise test.

https://github.com/rurban/smhasher/blob/9ca488454eda4dae9ac795147696135e5cdf7dbe/doc/xxh128.txt#L816

But `'PerlinNoise'` test contains new PerlinNoiseAV (`AV variant`) test.

https://github.com/rurban/smhasher/blob/9ca488454eda4dae9ac795147696135e5cdf7dbe/doc/xxh128.txt#L859

**Comment by jemfinch** [comment]
I dug into this and found a few things:

0. To answer your previous question, the problem is in PerlinNoiseAV: https://github.com/rurban/smhasher/blob/master/KeysetTest.h#L358
```
Testing collisions (128-bit) - Expected    0.0, actual  24116 (1711423593602593105674976100352.00x) (24116)
Testing collisions (high 64-bit) - Expected          0.0, actual  24116 (92776458911.34x) (24116) !!!!!
Testing collisions (high 32-bit) - Expected       1116.2, actual  25260 (22.63x) (24144) !!!!!
Testing collisions (high 25-37 bits) - Worst is 37 bits: 24162/34 (692.56x) !!!!!
Testing collisions (low  64-bit) - Expected          0.0, actual  24116 (92776458911.34x) (24116) !!!!!
Testing collisions (low  32-bit) - Expected       1116.2, actual  25187 (22.57x) (24071) !!!!!
Testing collisions (low  25-37 bits) - Worst is 37 bits: 24153/34 (692.30x) !!!!!
```
1. That keyset consists entirely of messages with length of 16 through 38 bytes, consisting almost entirely of zeroes, with one nonzero 32-bit value at some location in the message.
2. The collisions with xxh128 happen entirely on 16-byte messages, so the problem is in XXH3_len_9to16_128b.
3. Visually, it appears that the collisions happen almost entirely when both the seed and the key are values with very few (2-4) bits set.  For instance, here's one of the hash collisions, printed as bitsets:
```
hash=11111111111101000100000010101111001000111110101100110000001100001000010010111000001100101010110000111110000101100001111101110011
  seed=00000000000000000000000100010000 key=00000001000000010000000000010000
  seed=00000000000000000000000100010001 key=00000001000000010000000000010001
  seed=00000001000000000000000100010000 key=00000000000000010000000000010000
  seed=00000001000000000000000100010001 key=00000000000000010000000000010001
```
4. A simple fix that doesn't seem to impact performance (at least on my my machine) is to improve the seed: `seed = XXH3_mul128_fold64(seed, PRIME64_3);`  With this line added at the beginning of XXH3_len_9to16_128b, PerlinNoiseAV passes:
```
Testing AV variant, 128 count with 4 spacing, 4-12:
Testing collisions (128-bit) - Expected    0.0, actual      0 (0.00x)
Testing collisions (high 64-bit) - Expected          0.0, actual      0 (0.00x)
Testing collisions (high 32-bit) - Expected       1116.2, actual   1137 (1.02x) (21)
Testing collisions (high 25-37 bits) - Worst is 37 bits: 48/34 (1.38x)
Testing collisions (low  64-bit) - Expected          0.0, actual      0 (0.00x)
Testing collisions (low  32-bit) - Expected       1116.2, actual   1109 (0.99x) (-7)
Testing collisions (low  25-37 bits) - Worst is 35 bits: 140/139 (1.00x)
```
5. In case any other length-specific functions struggle with seeds that don't have very many bits set, you can lift that `XXH3_mul128_fold64(seed, PRIME64_3)` all way up to XXH128 without (apparently) impacting performance on small messages or performance on SMHasher (apart from fixing this bug, of course).  And, of course, callers can simply do this on their side without changing XXH128 at all.

**Comment by avaneev** [comment]
Yes, a reasonable fix (wyhash did that as well recently) is to add an additional "input seed" shuffling round.

**Comment by Cyan4973** [comment]
This test has nothing to do with Perlin noise.

The original PerlinNoise test (64b7d10) sweeps a 2D grid of (input, seed) coordinates and checks that the resulting hashes are well distributed. This is a textbook scenario, in application in real production software, I even reviewed paper on this topic.

PerlinNoiseAV (d3eeb20) does not do that: it builds near-empty messages (a single 4-byte word in 256 zero bytes) and sweeps sparse seeds/keys (7 bits spaced by 4) over the 16–38 byte range. There is no real world application that legitimately employ this pattern, and it is targeting precisely the 17–128 short path. Combined with how the seed is folded into `XXH3_mix16B` (secret ± seed), this is a targeted search over (input, seed) structures chosen to exploit the multiplicative formula, not a distribution test. In short, an adversarial pattern, targeting a non-cryptographic hash.

xxHash's contract is good dispersion of the hashes produced from normal data. 
It makes no claim against patterns specifically crafted to attack a published non-cryptographic formula.
The seed is explicitly not a security boundary, and adversarial inputs like these are outside that contract by design.

**Comment by avaneev** [comment]
I agree, this is more a nit-picking than an issue. However, this test does prove that universality claims (if there are or will be any) cannot be applied to xxhash without thorough overhaul.

I would more worry about SMHasher3 reports - these are real systemic issues. https://gitlab.com/fwojcik/smhasher3/-/tree/main/results?ref_type=heads

**Comment by avaneev** [comment]
I would also like to make you aware that the current rise of AI model usage created a nervous landscape for hash-function developers. AIs can dig deep. https://github.com/avaneev/komihash/issues/23

**Comment by thomasahle** [comment]
@avaneev, I reported some "high collision probability pairs" here as well: https://github.com/Cyan4973/xxHash/issues/1127#issuecomment-5741976219
Do you think the fact that AI makes it so easy to find these kinds of collisions changes anything in the calculus of whether universal hashes are worth it?

**Comment by avaneev** [comment]
@thomasahle This would be a real find for cryptographic hashes, but not for non-cryptographic hashes. If AI can find such for common cryptographic hash functions, that may reshape cryptography drastically.
